### ISSUE TITLE
Problem with APIConnector.upload_flow (?)

### ISSUE THREAD

**Issue by PG-TUe** [openml/python#28]
I want to upload a flow description programmatically. I have a [flow description](https://github.com/openml/python/blob/feature/script/RandomForestClassifier_flow.xml) which I want to upload using the [APIConnector.upload_flow](https://github.com/openml/python/blob/feature/script/openml/apiconnector.py#L952) method. 

I currently call the method with a description 'test description', and the absolute path to the file. The return value I get is
```
connector.upload_flow('test description', abs_path)
>>>(200, <Response [200]>)
```
with the repsonse.content containing
```
<oml:error xmlns:oml="http://openml.org/openml">
\t<oml:code>163</oml:code>
\t<oml:message>Problem validating uploaded description file</oml:message>
\t\t<oml:additional_information>Start tag expected, \'&amp;lt;\' not found. 
XML does not correspond to XSD schema. </oml:additional_information>
\t</oml:error>
```
It seems like everything works fine, but my file is not correct. But I don't see what is wrong with the file. Oddly enough, I *can* upload it through the [API docs](http://www.openml.org/api_docs/#!/flow/post_flow). I will actually see it show up under my flows on the openml website, and as a response I get the new flow id.

Anyone got ideas?

**Comment by joaquinvanschoren** [comment]
It looks like your file is being HTML encoded. \'&amp;lt;\' is HTML
encoding for '<', your opening tag.
Can you track down where you are actually uploading the file and why it is
using HTML encoding?

Cheers,
Joaquin

**Comment by PG-TUe** [comment]
Looks like it is within [this](https://github.com/openml/python/blob/master/openml/apiconnector.py#L871) external call. For what it is worth, the method 'post' is called with the following arguments (in order):
```
http://api_new.openml.org/v1//flow/
{'description': 'test description', 'api_key': my_api_key}
{'source': <_io.BufferedReader name=abs_path>}
```
where my_api_key is my actual api key (as string), and abs_path the actual absolute path to the file (as string).

**Comment by joaquinvanschoren** [comment]
Maybe check this out:
http://docs.python-requests.org/en/latest/user/quickstart/
Especially the part about multipart-encoded files. Are you reading the
files in binary mode?

**Comment by PG-TUe** [comment]
All files sent are read in binary by the apiconnector ([here](https://github.com/openml/python/blob/master/openml/apiconnector.py#L866)). Thanks for the reference, I'll try to call the methods based on the quickstart on my own, see if that helps.

**Comment by PG-TUe** [comment]
Doing it manually gets me the response:
```
<oml:error xmlns:oml="http://openml.org/openml">
\t<oml:code>102</oml:code>
\t<oml:message>Please provide session_hash</oml:messag
```
I tried to see if/where the APIConnector creates a session (or has the hash), but I couldn't find it (yet?).

**Comment by joaquinvanschoren** [comment]
apparently you add it to the data object, from the code:

def _read_url(self, url, data=None, file_dictionary=None): data['api_key'] =
self.config.get('FAKE_SECTION', 'apikey')

**Comment by PG-TUe** [comment]
My bad. I did add the hash to the data, but erroneously added it with the key 'apikey' (lacking the _ character). The response I now get is exactly the same as I got initially (so it would still be html encoded).

**Comment by PG-TUe** [comment]
I just uploaded a run through the APIConnector, which worked just fine. This would is odd, as it uses exactly the same methods internally (just a different url it posts the files to). Could this indicate that there is something wrong the my xml file after all? If so, how would it be possible that the api docs do not catch it? Is it handled differently internally?

**Comment by joaquinvanschoren** [comment]
Can you send me the flow XML? Did you check it with an editor like vi that
doesn't interpret HTML?
Maybe the api_docs do something special internally to correct bad files.

**Comment by PG-TUe** [comment]
The [flow description](https://github.com/openml/python/blob/feature/script/RandomForestClassifier_flow.xml) (as linked in the original post).

**Comment by PG-TUe** [comment]
Oh, I just noticed you've been replying by e-mail. I imagine the linking might not work correctly? Here is the full link: https://github.com/openml/python/blob/feature/script/RandomForestClassifier_flow.xml

**Comment by joaquinvanschoren** [comment]
The file looks fine. It must be something on the python side.
Are you using
file_elements[key] = open(path, 'rb')
to read the file?
Did you check whether it is read correctly? Maybe you are uploading an
empty file?

Reading the trace again, I do not think it is an encoding problem: the
\'&amp;lt;\' thing is returned in the response, so it may just be encoded
after receiving the response, when the error (i.e. the server did not see a
'<' character when reading the file) has already happened.

**Comment by PG-TUe** [comment]
Alright, I just uploaded the code I'm using in [this commit](https://github.com/openml/python/commit/13eac2ab71a62de16139b439c7eb816480229549), specifically [these lines](https://github.com/openml/python/blob/feature/script/openml/autorun.py#L139-L147) are what's doing the work right now.
Because I call upload_flow, eventually [file_elements[key] = open(path, 'rb')](https://github.com/openml/python/blob/feature/script/openml/apiconnector.py#L887) does get called. I even just printed the file there (to the console), and everything is in there at the time the file is read.

I also tried removing the newline characters, and removing the first line (which declares the encoding), neither helped.

**Comment by joaquinvanschoren** [comment]
Thanks. Can you give a few lines of code you use to test this (and that
generates that error)?
@zardaloop: we are able to upload datasets, right? Including XML
description?

**Comment by zardaloop** [comment]
Yes as far as I know it should work https://github.com/openml/python/blob/master/openml/apiconnector.py#L917
Cheers

P.s:  @joaquinvanschoren Thanks for sorting out the issue with the dataset yesterday, I have tested it and it worked fine.

**Comment by PG-TUe** [comment]
I think the minimum is this, having [this description](https://github.com/openml/python/blob/feature/script/RandomForestClassifier_flow.xml) somewhere locally :
```
from openml.apiconnector import APIConnector

key = KEY_HERE
connector = APIConnector(apikey=key)
abs_file_path = ABSOLUTE_PATH_TO_DESCRIPTION_FILE
return_code, response = connector.upload_flow('test', abs_file_path)
```

The return_code should be 200, and the response.content as described.

**Comment by joaquinvanschoren** [comment]
Sorry, I was snowed under. Are you still having problems with this?

**Comment by PG-TUe** [comment]
Yes. I have not yet found a solution (though to be fair, I have been busy with other work as well).

**Comment by joaquinvanschoren** [comment]
Alright, I got it to work.
1. You were uploading the XML description as the flow file (the file with the code), not the description. 
2. In the dataset upload code (which you copied), the description is passed as a string, not a file

Hence, instead of doing this:
```
return_code, response = connector.upload_flow('test', abs_file_path)
```

You need to do this:
```
flowxml = open(abs_file_path, 'rb').read()
return_code, response = connector.upload_flow(flowxml)
```

I committed a fix for the flow_upload method. You still need to call it as described above.

**Comment by PG-TUe** [comment]
I can hardly believe the solution was that simple. Thanks a lot!
One question about the form of implementation though (which is probably what got me confused in the first place):

When using upload_run to upload a run, you provide a dictionary with files:
```
files = {'predictions':path_1, 'description':path_2}
connector.upload_run(files)
```
While for a flow you would need to be done as you described: the description and extra file are separate parameters, and of the description one requires a file handler rather than a path.

I would suggest the following (for all upload functions):
 - using paths only, as the user shouldn't have to know we need to read it in binary in order to transfer the file.
 - every mandatory file gets its own parameter, and the optional files are grouped in a dictionary (though I'm not sure if I'm in favor of using a dictionary over parameters with defaults). This makes it clear what is mandatory and what isn't (eg. only a description for flow, but a description *and* predicitions for a run).

I'm willing to make these changes myself (if you agree to them), but might not be able to do them this weekend (quite busy for my web seminar deliverables ;) ).

**Comment by joaquinvanschoren** [comment]
I would avoid writing files to disk if not necessary.
Hence, I like the solution that the description is a separate parameter. As
a rule, it is also the only required parameter, as the server needs this to
know what to do with any other uploaded data.

Other parameters can be files or paths, but again avoid writing files to
disk. You can probably write a wrapper where you only need to enter paths
for external files?

**Comment by PG-TUe** [comment]
>I would avoid writing files to disk if not necessary.

Sure, that seems like a good thing to strife for. My example only used paths as that currently seemed like the only way to pass that information.

>As a rule, it is also the only required parameter

This is true for flows (and perhaps also tasks), but as far as I am aware it is not true for runs (they need a predictions-file) and datasets (they need the actual data). This is why I would extend make those explicitly necessary arguments in their respective upload methods.

For example, the current upload_run method just wants a dictionary of 'files'. Using help (help(APIConnector.upload_run)) gives no additional information. Going into the APIConnector.py file reveals that the dictionary actually requires a 'predictions' entry, otherwise the method will flat-out raise a ValueError. 

Then, if you actually call it with a dictionary that just contains the 'predictions' entry, it will actually attempt to upload it to the server, but get a response: Please provide run xml. In order to then actually upload the run succesfully, one has to add another entry to the files dictionary, the entry 'description' (note that the error message did not mention that it was a 'description' that was wanted - not only that, it has to be exactly spelled 'description' as the key, or it is not recognized as such).

The above example is the type of thing I would like to avoid - it severly increases the difficulty of using the API, in my opinion. The dictionary approach could be used together with a well documented functions. And while functions should be well documented anyway (and I assume it is simply a matter of time before they are), I personaly feel *all* required fields need to be explicit as parameters to better convey it. Even if that means that upload_run will have two mandatory parameters, and upload_flow only one.

>Other parameters can be files or paths, but again avoid writing files to disk. You can probably write a wrapper where you only need to enter paths for external files?

Would it be a good idea to allow the arguments to be either a string containing the xml or a path? And then just first check whether the string contains a path or an xml, and then process it as such?

**Comment by mfeurer** [comment]
Good to hear that you solved the problem. I'm not aware of how the other upload methods work, but all parameters which can be passed to the server should be a parameter to the respective API call. Furthermore, only parameters that are optional should have a default value. @PG-TUe it would be great if you can do the required changes. About the question file/xml. Is one of those a natural default? If yes, I would say it should have the name `upload_x()`, while the other has the name `upload_x_from_y()` and both call an internal method `_upload_x()`. What do you think of this?

**Comment by PG-TUe** [comment]
If I remember correctly, internally currently the APIConnector always sending a file, as we pass a file handler to the requests API. In that sense, a file seems more natural. But perhaps we can look into whether or not instead of providing a file, we can provide a string or other data.

What would also be nice is to be able to able to pass the OpenML objects themselves, eg. `upload_run(run)` where `run` is an OpenMLRun. But that does mean we then first have to check whether or not the run object has all the data required, etc. I think this means that it is more prone to being used incorrectly, and the only way we can give feedback to the user is through raising errors.

I will probably at least (as an intermediate step) try to make the different `upload_x` calls more similar to one-another, and make sure that mandatory files get their own mandatory parameter. I'm not sure when I'll get around to it, this week will still be quite busy.

**Comment by zardaloop** [comment]
Hi Guys, Sorry that I was not that involved over the past few months I have been super busy with work and study.
Is this something you guys want me to fix ? I had a quick look at the Joaquin fix and it seems reasonable.  
@PG-TUe  thanks for the feedback. Few issues I have noticed here, first of all for the upload_flow method it seems the file and description are both required so the predefine path being set to None is wrong. @joaquinvanschoren  any idea why is that?? It seems as long as you upload a text file it works no matter what, what is that file exactly ??

Also regarding the upload_run method, @PG-TUe  I agree with you we should only pass the file_path and the description as a text not a dictionary, to be honest I am not sure why I have done that but that was the last thing I was implementing during the hackethon, no excuse for that so I will fix that hopefully this weekend. 
Is there anything else need to be fixed ?

**Comment by joaquinvanschoren** [comment]
For flow uploads, the API does not require a file to be uploaded, it is
only optional. http://www.openml.org/api_docs/#!/flow/post_flow
That file can contain the code for the flow (e.g. a Python or R file), but
this doesn't make sense if you use algorithms from a library. Hence it is
not required.

I slightly lost track of the latest developments in upload_run but if I
understand correctly it is now implemented as a standalone script. It would
be great if this could be integrated in the core module soon.

@PG-TUe, did you have any open problems that need to be fixed?

Cheers,
Joaquin

**Comment by PG-TUe** [comment]
Uh, I guess this kind of rounds up the discussion, so I will close it.

@zardaloop (as stated by joaquin:) for a flow upload there are two files: a description, which is mandatory, and an optional file which contains the code of the flow. I was planning to rewrite it myself in the coming week (I'm wrapping up one assignment for school, and then I'm back to finishing openml_run). So don't feel obligated to fix it this weekend (though if you do, I'll just be pulling in your changes instead ;) )
@joaquinvanschoren Everything seems clear to me now, so I think I'll be able to wrap it up either in the coming week or the schoolbreak that follows it.

**Comment by zardaloop** [comment]
@joaquinvanschoren  that's what I thought, it also doesn't make sense to me either to upload any file when we are uploading the flow. But I have tested that without uploading the file and it failed.
Regarding the development on upload_run I have implemented that function before, but based on @PG-TUe  I have simplified the implementation of it on a new branch. Apart from that I have not change any code since the hackethon and the only thing is changed is the hotfix you have added couple of weeks ago.  
I am not going to merge that until it is fully tested, and preferably . I will look into that over the weekend and I will also investigate to see what is wrong with the implementation of the upload_flow. For now @mfeurer simply ignore the branch and the pullrequest I have created until I fully test that.

**Comment by PG-TUe** [comment]
@zardaloop Sorry for the late reaction, I somehow missed this. I have been working on the script at the feature branch feature/script, the particular file is [this one](https://github.com/openml/python/blob/feature/script/openml/autorun.py), though I there have been a few changes to apiconnector.py too.

My intention is to finish it over the weekend (with for now the restrictions on classification and n-fold estimations). I hope to add other functionality to it later (eg. Regression, Holdout).

**Comment by zardaloop** [comment]
@PG-TUe  Don't worry about it, I am exactly the same. I am also currently joggling too many stuff at the same time and very little time left for my phd studies and other side projects. 
Have a look at the branch I have created and I guess it would sense if you also make your commit to that branch and if we are happy with that then we will merge it into the develop branch and eventually we will merge it to the master branch.
Let me know how you getting on. 
Cheers :)

**Comment by PG-TUe** [comment]
@zardaloop To be frank, I have very little experience with Git merging et cetera. It is certainly something I plan to learn.. but I'm not sure how safe it is to try it. I suppose in the worst case we can rollback anyway :) 

I'll add/merge my work onto the feature/Upload_Fixes branch then, and keep working on my script from that branch? But wouldn't that delay the Upload_Fixes branch?

**Comment by joaquinvanschoren** [comment]
@zardaloop Is @PG-TUe supposed to push to the Upload_Fixes branch? I would
love these features to be in the master as soon as possible.

**Comment by zardaloop** [comment]
@joaquinvanschoren , well I have made some changes in that branch and I asked @PG-TUe to also commit his changes into the same branch and test it to see if it work. Then I will review it and will test again and if it also work for me I will merge it to develop if not I will fix it and then I will do that. After that I will make a release and merge it to the master. However I am not sure where he is with that.

**Comment by joaquinvanschoren** [comment]
Sounds good. @PG-TUe, do you have time to push your work (the run-related
functions) into that branch? Don't worry if you haven't done this before,
we can always fix things.

**Comment by PG-TUe** [comment]
Just to be clear; is it **just** my changes I made to pre-existing stuff (APIConnector, task, dataset, etc.) or *also* the openml_run script?

Either way I should be able to do that before the weekend, but would like to know what exactly is expected here :)

**Comment by joaquinvanschoren** [comment]
Definitely also the openml_run components :).
It would be great if we would have a simple demo script (like the one in
your report) that does a complete run using the Python API.

**Comment by PG-TUe** [comment]
Just did it! It's on the Upload_Fixes branch now. Fixed a few issues I had because I currently use a different environment than I originally development one (the main one is that OrderedDict instantiation worked differently, resulting in an invalid xml file). Anyway.. let me know if you have any issues running it.

**Comment by joaquinvanschoren** [comment]
Awesome!
Should we merge apiconnector.py and autorun.py? Is there a good reason to
keep them separate?

**Comment by PG-TUe** [comment]
I would argue the other way around. I don't see a particular reason why they should be together (other than having it in a single file). They do very different things, which is why I would keep it separated.

Furthermore, this is a personal thing, but it feels like APIConnector is already getting very big. While Autorun.py is limited in size currently, if support gets extended to other validation procedures or regression et cetera - if that also happens to be done in the same file - it becomes very large as well. Having the two together seems unwieldy to me.

**Comment by joaquinvanschoren** [comment]
OK, it's just a bit weird that only the run part is now in a separate file.
Maybe we should indeed have separate files for data, tasks, flows, too...

Btw, looking at your example script, can't we move this inside
openml.upload_run so that the user doesn't have to worry about it?
prediction_abspath = os.path.abspath(prediction_path)
description_abspath = os.path.abspath(description_path)

Also, maybe it should return a json object instead of the xml object? Then
the user doesn't need to worry about the xml parsing either.

**Comment by PG-TUe** [comment]
Regarding the absolute paths, I don't think I should do it in the example script either. The best solution (I think) would be to do it where you actually read the file. I opened [an issue](https://github.com/openml/python/issues/31) for this. Though if that is not a solution, I would move it inside openml_run, yes. :)

Yes, ideally the end-user should be able to modify at least the run description (and maybe also the predictions, if they want to add a description note). That would be easier if it was not written to file yet. Either by passing back the ordered dict, or, maybe better, to somehow make it possible to generate a run description from an OpenMLRun task. That way, we could just pass an OpenMLRun object to upload_run, as well as return a OpenMLRun object from openml_run.

I just wasn't sure what the best way would be, and I don't have time to implement that quickly enough if we want to process the feature/Upload_Fixes branch - so I opted to solve it like this for now.

**Comment by PG-TUe** [comment]
Oh and
>OK, it's just a bit weird that only the run part is now in a separate file.
Maybe we should indeed have separate files for data, tasks, flows, too...

That might be 'correct'. But there is something to be said for having it separated from openml_run either way. Openml_run's task is to set up a machine learning experiment, not to have API interactions per se. The only place it does that right now, is when making sure a flow exists.. and I'm not really sure it belongs there (other than that it is required.).

**Comment by joaquinvanschoren** [comment]
I see, you mean that this is tied to sklearn, not part of a general API
connector

**Comment by PG-TUe** [comment]
Exactly :)

**Comment by zardaloop** [comment]
I just had a chance to look at the code. Many thanks @PG-TUe for your effort. Also @joaquinvanschoren  thanks for adding the Jupyter notebook.
I would like to nag a little bit about the fact that the branch Upload_Fixes straight away was merged into master branch  :smile:   .  We should make sure as a minimum we are following the Gitflow pattern. 

I have rebased develop branch on master, created a new branch and made my changes. Then I have merged that branch back to develop and created release branch 0.1.3 and finally merged that branch to both master and develop.

I have moved the example code for sklearn into the examples folder and made some changes to the tests. So now at least most of them are passing. 

I am surprise why these tests were not fixed before :open_mouth: .
Lets make sure from now on we will never merge something that is not tested into master branch no matter how vital it is :wink: . There might be still some tests that are failing but I did as much as I could for now.

I have also cleaned up a little bit of a code in upload__flow to make it easier to read.

If there is any issue with what we have in master branch and we want to fix it quickly and it is NOT a new feature but is ONLY a simple bug fix, we should create a hotfix branch.

These little things will help us to make the code more manageable and easier to maintain. Otherwise it will be really hard to maintain as it grow.

@PG-TUe When you get a chance could you please test the api and if you are happy delete the Uload_Fixes branch. Also I can see you have two other branches feature/iterate-splits-of-task  and feature/script . What are these branches ? Are you still working on them? 
if you don't need them anymore please make sure to close them or if they are something that need to be merged into the develop branch and you are done working on them, please go ahead and make a pull request for develop branch so either myself,  @joaquinvanschoren  or @mfeurer  will review that and will merge it into the develop so we can include it into the next release of the application.

**Comment by PG-TUe** [comment]
@zardaloop 
I will delete the feature/script branch, as it had been merged into upload_fixes.
The iterate-splits-of-task branch was not integrated, I wasn't aware it was not tagged for merger yet, as it had been ready for a few weeks now. It basically makes it possible to retrieve splits (repeat/folds) directly from a task instead of through downloading splits first.

**Comment by joaquinvanschoren** [comment]
Sorry, this is my mistake. I wanted to push my Notebook example to
Upload-fixes, but pushed to master by mistake. What I did not notice is
that I actually merged the code while doing this!! Terribly sorry about
this.

I'll run the tests again and check whether everything still works.

Cheers,
Joaquin

**Comment by zardaloop** [comment]
That's fine, it is not a big deal. I don't think too many people are using the library yet. However if that happens in the future then that might break a lot of think for people and dealing with such issue will be very difficult. 
I have merged everything and cleaned up all the branches and release version 0.1.4 . 
If there is anything that is not working lets make an issue for that first and then we can judge if is a new feature or a bug. 
For now I have assumed all these new stuff are more of a bugs that a feature. The reason behind that is if is a new feature then middle number bumps up but if is a bug the third number goes up. That is why I have bumped up third digit to 0.1.4. 

Since we don't have anything else to add I go ahead and release version 0.2.0 which includes the new features  and bug fixes that has been added recently.

Then from version 0.2.0 we will make sure to do things in the correct way :wink: .
